lib/full-site-editing/block-templates.php

```diff
@@ -0,0 +1,281 @@
+ *
+ * @return WP_Block_Template Template.
+ */
+function _gutenberg_build_template_result_from_post( $post ) {
+	$theme = get_the_terms( $post, 'wp_theme' )[0]->slug;
+
+	$template              = new WP_Block_Template();
+	$template->wp_id       = $post->ID;
+	$template->id          = $theme . '|' . $post->post_name;
+	$template->theme       = $theme;
+	$template->content     = $post->post_content;
+	$template->slug        = $post->post_name;
+	$template->is_custom   = true;
+	$template->type        = $post->post_type;
+	$template->description = $post->post_excerpt;
+	$template->title       = $post->post_title;
+	$template->status      = $post->post_status;
+
+	return $template;
+}
+
+/**
+ * Retrieves a list of unified template objects based on a query.
+ *
+ * @param array $query Query.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return array Templates.
+ */
+function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_template' ) {
+	$wp_query_args = array(
+		'post_status'    => array( 'auto-draft', 'draft', 'publish' ),
+		'post_type'      => $template_type,
+		'posts_per_page' => -1,
+		'no_found_rows'  => true,
+	);
+
+	if ( isset( $query['theme'] ) ) {
+		$wp_query_args['tax_query'] = array(
+			array(
+				'taxonomy' => 'wp_theme',
+				'field'    => 'slug',
+				'terms'    => $query['theme'],
+			),
+		);
+	}
+
+	if ( isset( $query['slug__in'] ) ) {
+		$wp_query_args['post_name__in'] = $query['slug__in'];
+	}
+
+	// This is only needed for the regular templates/template parts CPT listing and editor.
+	if ( isset( $query['wp_id'] ) ) {
+		$wp_query_args['p'] = $query['wp_id'];
+	} else {
+		$wp_query_args['post_status'] = 'publish';
+	}
+	$template_query = new WP_Query( $wp_query_args );
+	$query_result   = array();
+	foreach ( $template_query->get_posts() as $post ) {
+		$query_result[] = _gutenberg_build_template_result_from_post( $post );
+	}
+
+	if ( ! isset( $query['wp_id'] ) && ( ! isset( $query['theme'] ) || wp_get_theme()->get_stylesheet() === $query['theme'] ) ) {
+		$template_files = _gutenberg_get_template_files( $template_type );
+		foreach ( $template_files as $template_file ) {
+			$is_custom      = array_search(
+				wp_get_theme()->get_stylesheet() . '/' . $template_file['slug'],
+				array_column( $query_result, 'id' ),
+				true
+			);
+			$should_include = false === $is_custom && (
+				! isset( $query['slug__in'] ) || in_array( $template_file['slug'], $query['slug__in'], true )
+			);
+			if ( $should_include ) {
+				$query_result[] = _gutenberg_build_template_result_from_file( $template_file, $template_type );
+			}
+		}
+	}
+
+	return $query_result;
+}
+
+/**
+ * Retrieves a single unified template object using its id.
+ *
+ * @param array $id Template id.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return WP_Block_Template|null Template.
+ */
+function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
```

Comment:
How about using `$slug` instead of `$id` as an arg, and inferring `$theme` from `wp_get_theme()->get_stylesheet()` (as we're doing anyway on [R223](https://github.com/WordPress/gutenberg/pull/27910/files#diff-6ca30c26729f0fb3398ad5c6b4e82d1a83bfe5bd920e4cb6deca5646d9e578bdR223) already)?

Reply:
I expect this function to be used for any template including the ones from non active themes.

Reply:
But then, the fallback to block template files (R223 and below) won't work, will it?

Reply:
yes, the fallback to templates files is only applied to the current theme. I thought that we shouldn't be allowing access to these if the theme is not active. (may change later but this means the "definition of active theme" might change in FSE)

Reply:
Hmm, but then we'd have different behavior for the active theme (check DB, fall back to template files) vs inactive ones (check DB, don't fall back to files). Isn't there quite some risk for confusion? Might be worth considering limiting to the active theme only while we're still working out the behavior for inactive ones?

Reply:
> Isn't there quite some risk for confusion?

I don't think so, think about "removed themes" as well, where you can also have CPT entries and you could consider  using these as custom templates or template parts in your site. I do see the "inactive theme" as a "removed theme" personally where the files just are kept there. It shouldn't behave differently from any theme that is not used.

That said, I do believe that "themes" behavior and maybe intent is drastically changing in FSE, so I could see a world where there's more than "one" active theme or "none", just trying to start small. 

I don't think we can ignore the templales/tempate-parts that come from other themes entirely at the moment because if we do so, say you create a custom template part today, and you switch themes, you won't be able to use your custom template part in your new theme.

Reply:
> That said, I do believe that "themes" behavior and maybe intent is drastically changing in FSE, so I could see a world where there's more than "one" active theme or "none", just trying to start small.

Starting small is a very good strategy here IMO. While it makes perfect sense that the concept of themes (and what templates/parts can be used across a site) will get more blurred, I think it's important not to let this influence us too much while we're still working on a v1 (where the goal is more or less 'feature parity' with WP's 'old' (PHP) template resolution). Otherwise, if we try to anticipate that future too much, there's a risk that it'll be harder to communicate that to other stakeholders who might not share the same mental model, to agree on some architectural decisions, and to make swift progress.

> I don't think we can ignore the templales/tempate-parts that come from other themes entirely at the moment because if we do so, say you create a custom template part today, and you switch themes, you won't be able to use your custom template part in your new theme.

In the vein of starting small, I'd be perfectly fine with that, at least for a v1 -- I even think it's better. Arguably, allowing to use templates/parts across themes could be considered an extra feature that deserves separate consideration and design. I'd go as far as to say that disallowing users from using templates/parts across themes might even be better UX (tighter coupling of templates/parts to a theme; a theme continues to be defined as a set of templates/parts). In any case, it seems like a good candidate for an iterative approach: Starting with a tighter constraint (which typically allows for a simpler implementation), get that implementation right (and cover with unit tests that reflect the needs); loosen that constraint in a subsequent iteration, and modify the implementation and tests accordingly.

(This might read bit more intense that it was meant to be; it just got a bit more verbose as I was trying to discuss different aspects/angles :blush: )

Reply:
I'm ok with trying that. Let's see.

Reply:
> Per the discussion we had we @ockham here #27910 (comment) I removed the fact that you can access templates and template parts (custom or not) when you switch themes.

Going down that route, would it make sense to update the template-part block to no longer request per the `theme` block attribute and to use the equivalent of `wp_get_theme()->get_stylesheet()` there as well?  That should fix the issue with template parts not loading for themes in nested directories, and I'm thinking without side effects (?) now that we cannot access either of them after theme switch.

Reply:
I don't think we should be swapping the theme from the id with wp_get_theme()->get_stylesheet() if that's  what  you're suggesting but I do think we should just return 404 if the theme is different.

Reply:
> I don't think we should be swapping the theme from the id with wp_get_theme()->get_stylesheet() if that's what you're suggesting

Why not?  If we are limiting it to the current theme, then the `theme` value added to the template part block in markup is entirely irrelevant (its always going to correspond to the active theme?) and `wp_get_theme()->get_stylesheet()` would accurately match what is being seen by the back end.

> I do think we should just return 404 if the theme is different.

Im not saying the theme is different.  Im saying the backend is going to hold the object as `pub/twenty-twenty-one-blocks|header` while the block is going to be requesting `twenty-twenty-one-blocks|header`.  It would SEEM that the theme is different, but its a false negative.  If we are not allowing access to templates/template-parts when we switch themes, there should be no reason for themes to even supply a `theme` attribute in the markup.

Reply:
> Why not?

Because, If I edit a template part A for theme1, I shouldn't be using it as a replacement of template part A  of theme2. These could be used in very different contexts and might not fit in the templates of the new theme.

Reply:
> If I edit a template part A for theme1, I shouldn't be using it as a replacement of template part A of theme2.

Im a bit confused.  How would this happen?  If you edit template part A for theme1, its saved with the id for theme1.  If you load theme 2, it wouldn't replace anything?

Reply:
I guess I was confused too  yes, It won't you're right but there's still a potential of confusion.

 If your theme is "theme1" and you try too fetch "theme2//something", it will return "theme1//something" which IMO is wrong.

Reply:
> If your theme is "theme1" and you try too fetch "theme2//something", it will return "theme1//something" which IMO is wrong.

I would agree that would be wrong wrong, but I thought you were going with not allowing access to templates/template-parts after switching themes?

Reply:
yes, so that's why I was suggesting a 404

Reply:
> If your theme is "theme1" and you try too fetch "theme2//something"

So i guess, where would we be trying to fetch it by a non-active theme?

Reply:
I don't know, it's a REST API :) so while in the UI, I expect it shouldn't exist, I'm just thinking in terms of correctness of the REST API.

Reply:
> I don't know, it's a REST API :) so while in the UI, I expect it shouldn't exist, I'm just thinking in terms of correctness of the REST API.

True, that makes sense... but it seems like a separate concern.  Couldn't we both:

* use the equivalent of `wp_get_theme()->get_stylesheet()` from the template-part block (fixing the nested theme directory issues)
* Update the endpoint to return 404 if the the theme in the id doesn't match the active theme?

They seem unrelated, and doing both would make sense?

Reply:
> use the equivalent of wp_get_theme()->get_stylesheet() from the template-part block (fixing the nested theme directory issues)

I'm not sure I understand this point. Could you elaborate?

Reply:
> I'm not sure I understand this point. Could you elaborate?

The template part block relies on the `theme` block attribute to create the id for the request to get the template part.  The `theme` block attribute is currently supplied by markup.  Using this to create the `id` to request is a fallacy because the block markup can never accurately account for where the theme will be installed.  Regardless of whether or not the theme is installed in a nested directory, this value will be the same.  While on the backend `wp_get_theme()->get_stylesheet()` is used, which includes the directory path in the themes value.  Thus no matter how the markup is written, themes will be 'broken' and template parts will not load if the theme happens to be installed in or not in a certain file structure.

Since we are limiting access to templates and template parts from other themes, this `theme` block attribute is now completely irrelevant.  It always corresponds to the active theme.  So instead of using the `theme` attribute supplied by markup and requiring that the theme is installed without being in a nested directory for the template part to be found, we can just use the equivalent of `wp_get_theme()->get_stylesheet()` instead such that the values on both sides of the request are consistent with including the file path and themes work regardless of where they are installed.

Reply:
> Regardless of whether or not the theme is installed in a nested directory, this value will be the same

I agree that right now we have the requirement that the theme should be basically using the same "directory" as its "id".  Whether that is a subdirectory or not and it's problem because themes can be copy/pasted into sub folders.

> Since we are limiting access to templates and template parts from other themes, this theme block attribute is now completely irrelevant. It always corresponds to the active theme

This solution for me is a temporary hack to the problem though and not a real solution, it just assumes a reality of the moment: templates can't  be used across themes, but we know for certain that we want that at least for custom created templates and template parts, so I'm not sure we should basing the solution to the "folder" problem on the fact that currently we ignore templates and template parts from other themes.

Reply:
I've been thinking more here, and actually, the solution I'd propose is very similar to yours but instead of it  being backend based, I think it should be frontend based, meaning:

 - template parts referenced in theme files shouldn't be using "theme" attribute
 - when finding a template part block without theme property in the editor, fallback to the active theme,
 - when a template is saved for the first time, explicitly assign the "theme" attribute for its template parts.
 - when a template is exported into the  zip, remove the "theme" attribute from its template parts. (When we'll be having cross theme  references, we'd need more than just removing the "theme", we might also have to rename "slugs" for the the template parts coming from other themes to avoid conflicts)

Reply:
Step 2 and 3 could potentially be merged and absorbed by the REST API to retrieve file templates (automatically inject the theme property when assigning the "content" property in `_gutenberg_build_template_result_from_file` function).

Reply:
> This solution for me is a temporary hack to the problem though and not a real solution, it just assumes a reality of the moment: templates can't be used across themes, but we know for certain that we want that at least for custom created templates and template parts, so I'm not sure we should basing the solution to the "folder" problem on the fact that currently we ignore templates and template parts from other themes.

Its a start that we can iterate from.  And any solution to the folder problem seems like we cannot rely on a value supplied by markup, so we should probably try to phase that out sooner rather than later.

Once we get into allowing custom templates and template parts from other themes we will need to address edge cases.  An idea would be to add the `theme` block attribute once the template or template part is becomes customized.  So:
* A template part supplied by a theme has no `theme` block attribute.  Once the template or template part is customized, we add this attribute to the template part blocks in its correct form.
* If the `theme` block attribute is not present on the template part block, we assume current theme from `wp_get_theme()->get_stylesheet()` and use that to build the id to request.  Otherwise, we use the `theme` block attribute to build the id for the request, and since we applied it with the correct value through Gutenberg it should work regardless of how the themes are installed.

There may be other things to address, but I think no matter how we go about it we need to stop relying on the `theme` attribute supplied by markup. 🤔

Reply:
> Its just one idea and there may be other things to address, but I think no matter how we go about it we need to stop relying on the theme attribute supplied by markup. 🤔

I agree with this and actually I think this is the simplest and most promising solution 
https://github.com/WordPress/gutenberg/pull/27910#discussion_r554204327

Reply:
> I agree with this and actually I think this is the simplest and most promising solution #27910 (comment)

I missed those 2 comments come in before I posted my last response.  😆   But yes!  That makes sense to me and seems like a good way to go about it.

Reply:
> This solution for me is a temporary hack to the problem though and not a real solution, it just assumes a reality of the moment: templates can't be used across themes, but we know for certain that we want that at least for custom created templates and template parts, so I'm not sure we should basing the solution to the "folder" problem on the fact that currently we ignore templates and template parts from other themes.

I think the `theme` attribute could be eliminated altogether (I proposed it at #26950) and here's why.

The `wp_theme` taxonomy enables multiple themes to be assigned to a post. If a user wants to switch themes but keep one of their old templates, this template plus any template parts it includes could be doubly assigned to the new theme. The user would get one chance (upon theme switch) to carry over whatever they want into the new theme, and once this is done, there would be no more need to reference the old theme on the block-level.

Reply:
Another benefit of this alternative approach is it allows for consistency between templates and template parts when it comes to using them "across themes." The block attribute can't be implemented for templates, since there is no "template" block, but the taxonomy term can work in the same way for both.

Reply:
> Step 2 and 3 could potentially be merged and absorbed by the REST API to retrieve file templates (automatically inject the theme property when assigning the "content" property in _gutenberg_build_template_result_from_file function).

@youknowriad Is this what you meant? https://github.com/WordPress/gutenberg/pull/28088/files. I wrote a quick and rough commit to inject a theme attribute into template content, and I'm happy to continue building on this when I wake up tomorrow.